tool/src/callback_dictionary.ts

```diff
@@ -1,28 +1,130 @@
 import hljs from "highlight.js";
 import mathematica from "highlight.js/lib/languages/mathematica";
+import javascript from "highlight.js/lib/languages/javascript";
+import typescript from "highlight.js/lib/languages/typescript";
+import php from "highlight.js/lib/languages/php";
+import ruby from "highlight.js/lib/languages/ruby";
 
 import { expand, StringObject } from "./util.js"
+import _ from "lodash";
 
-export const callbackDictionary = new Map<string, string>([
-  [
-    hljs.END_SAME_AS_BEGIN({})["on:begin"]!.toString(),
-    "endSameAsBeginOnBegin"
-  ],
-  [
-    hljs.END_SAME_AS_BEGIN({})["on:end"]!.toString(),
-    "endSameAsBeginOnEnd"
-  ],
+const commonCallbacks = new Map<string, string>(
   [
-    hljs.SHEBANG()["on:begin"]!.toString(),
-    "shebangOnBegin",
-  ],
+    [
+      hljs.END_SAME_AS_BEGIN({})["on:begin"]!.toString(),
+      "endSameAsBeginOnBegin"
+    ],
+    [
+      hljs.END_SAME_AS_BEGIN({})["on:end"]!.toString(),
+      "endSameAsBeginOnEnd"
+    ],
+    [
+      hljs.SHEBANG()["on:begin"]!.toString(),
+      "shebangOnBegin",
+    ],
+  ]
+);
+
+export const callbackDictionary = new Map<string, string>([
+  ...commonCallbacks,
 
   ...expand([
-    mathematica,
+    javascript(hljs),
+    mathematica(hljs),
+    php(hljs),
+    ruby(hljs),
+    typescript(hljs),
   ].map(language => getLanguageCallbacks(language))),
 ]);
 
 function getLanguageCallbacks(language: StringObject<any>): [string, string][] {
-  // TODO: Extract all callbacks recursively.
-  return [["a", "b"]];
+  let result: [string, string][] = [];
+
+  const addResult = (k: string, v: string) => {
+    if (!commonCallbacks.has(v.toString())) {
+      result.push([v.toString(), k]);
+    }
+  };
+
+  let objects = <StringObject<any>>[];
+  let languageName: string = language.name;
+  let currentName: string[] = ['k'];
+  if (languageName !== undefined) {
+    currentName.push(languageName.toLowerCase());
```

Comment:
Throw an exception if there is no name, and abort the entire porting.

Reply:
Turns out, PHP doesn't have a `name` field...

It does have a callback on the main branch, but not in the latest version.